apps/frontend/core-dapp/src/stores/entities/UniswapV3PoolHistoricalData.ts

```diff
@@ -0,0 +1,212 @@
+import { makeAutoObservable, reaction, runInAction } from 'mobx'
+import { SEC_IN_MS } from 'prepo-constants'
+import { MarketEntity } from './MarketEntity'
+import { RootStore } from '../RootStore'
+import {
+  ChartTimeframe,
+  FormatPoolsHistoricalDatasOptions,
+  HistoricalDataQueryType,
+  PoolsDayDatas,
+  PoolsHourDatas,
+  SeparatedMarketHistoryData,
+} from '../../types/market.types'
+import { DateRange } from '../../types/general.types'
+import {
+  formatMarketHistoricalData,
+  normalizePoolsDayDatas,
+  normalizePoolsHourDatas,
+  syncPoolsData,
+} from '../../utils/market-utils'
+import { PROJECT_START_TIMESTAMP } from '../../lib/constants'
+import {
+  getDateRangeFromDays,
+  getDateRangeFromHours,
+  getDaysFromDateRange,
+  getEndOfHour,
+  getHoursFromDateRange,
+  getStartOfHour,
+  getUTCEndOfDay,
+  getUTCStartOfDay,
+} from '../../utils/date-utils'
+import { UNISWAP_MAX_DATAPOINTS } from '../graphs/UniswapV3GraphStore'
+
+const timeframeMap = {
+  [ChartTimeframe.DAY]: 24,
+  [ChartTimeframe.WEEK]: 168,
+  [ChartTimeframe.MONTH]: 720,
+  [ChartTimeframe.YEAR]: 8760,
+  [ChartTimeframe.ALL]: UNISWAP_MAX_DATAPOINTS * 24,
+}
+
+const getProjectStartedHours = (): number => {
+  const now = new Date().getTime()
+  const projectStartTimeInMs = PROJECT_START_TIMESTAMP * SEC_IN_MS
+  return getHoursFromDateRange({ endTimeInMs: now, startTimeInMs: projectStartTimeInMs })
+}
+
+/**
+ * Handles caching historical data of a market:
+ * The UniswapV3GraphStore exposes 2 functions for each timeframe - one to get historical data and one to get latest data
+ * It is designed this way to allow fetching the latest data every block without having to refetch the entire historical data
+ * Because fetching the entire historical data is both computationally expensive and slow
+ *
+ * This store handles caching historical data, so we don't have to refetch them when timeframe is switched,
+ * and intelligently subscribe to latest data of the selected timeframe
+ */
+export class UniswapV3PoolHistoricalData {
+  cachedTimeframe = ChartTimeframe.DAY
+  cachedHistoricalData?: SeparatedMarketHistoryData
+  constructor(private root: RootStore, public market: MarketEntity) {
```

Comment:
```suggestion
  constructor(private root: RootStore, public market: BaseMarketEntity) {
```

Should probably use `BaseMarketEntity`? This class doesn't reference any property specific to `MarketEntity`

Reply:
yeap good point